### PR TITLE
CDAP-2016: CLI support for starting programs in debug mode

### PR DESCRIPTION
**Pull request by jwang47**
https://issues.cask.co/browse/CDAP-2016
http://builds.cask.co/browse/CDAP-DUT1883-4

### CODE REVIEW

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/StartDebugProgramCommand.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright © 2014 Cask Data, Inc.
```
Comment (awholegunch) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/StartDebugProgramCommand.java]: 2015!

File: cdap-client/src/main/java/co/cask/cdap/client/ProgramClient.java [hunk 2]
```diff
@@ -92,15 +93,22 @@ public ProgramClient(ClientConfig config) {
    * @throws ProgramNotFoundException if the program with the specified name could not be found
```
Comment (awholegunch) [inline cdap-client/src/main/java/co/cask/cdap/client/ProgramClient.java]: missing the debug parameter in the Javadocs. Goes above runtimeArgs; can be copied from the method after this one.

File: cdap-client/src/main/java/co/cask/cdap/client/ProgramClient.java [hunk 3]
```diff
@@ -112,6 +120,23 @@ public void start(String appId, ProgramType programType, String programName, Map
    * @param appId ID of the application that the program belongs to
```
Comment (awholegunch) [inline cdap-client/src/main/java/co/cask/cdap/client/ProgramClient.java]: Should change the method description to indicate why this method has been added. How about:
"Starts a program, giving debug mode, using the stored runtime arguments."

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/StartDebugProgramCommand.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    String[] programIdParts = arguments.get(elementType.getArgumentName().toString()).split("\\.");
+    if (programIdParts.length < 2) {
+      throw new CommandInputError(this);
+    }
+
+    String appId = programIdParts[0];
+    String programId = programIdParts[1];
+
+    String runtimeArgsString = arguments.get(ArgumentName.RUNTIME_ARGS.toString(), "");
+    if (runtimeArgsString == null || runtimeArgsString.isEmpty()) {
```
Comment (sreevatsanraman) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/StartDebugProgramCommand.java]: I don't think we need to fetch the run time args and pass it. The preferences API would resolve the stored runtime arguments and pass it to the program. Can you open a JIRA for fixing this in 3.1, there are other places where this needs to change as well and not only for debug,

**Comment by sreevatsanraman** [comment]
Open a JIRA for the issue mentioned and merge. 

LGTM :+1:

**Pull request merged**